### PR TITLE
[cxx-interop] Allow initializing `std::function` from Swift closures

### PR DESCRIPTION
**Pull request by egorzhdan**
This adds a Swift initializer to instantiations of `std::function` that accepts a Swift closure with `@convention(c)`.

rdar://103979602

### CODE REVIEW

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

File: test/Interop/Cxx/stdlib/use-std-function.swift [hunk 1]
```diff
@@ -5,25 +5,50 @@
...
+  expectEqual(0, f2(0))
+  expectEqual(246, f2(123))
+}
+
+StdFunctionTestSuite.test("FunctionIntToInt init from closure and pass as parameter") {
+  let res = invokeFunction(.init({ $0 * 2 }), 111)
+  expectEqual(222, res)
+}
+
+// FIXME: assertion for address-only closure params
```
[the remark this answers is not in the document]
Comment (egorzhdan) [reply test/Interop/Cxx/stdlib/use-std-function.swift]: This will be fixed by https://github.com/apple/swift/pull/73561

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
I'm going to merge this to unblock some other work, but I'm happy to address review comments post-merge.

File: test/Interop/Cxx/stdlib/use-std-function.swift [hunk 2]
```diff
@@ -9,25 +9,52 @@
...
-StdFunctionTestSuite.test("retrieve and pass back as parameter") {
+StdFunctionTestSuite.test("FunctionIntToInt retrieve and pass back as parameter") {
   let res = invokeFunction(getIdentityFunction(), 456)
   expectEqual(456, res)
 }
 
+#if !os(Windows) // FIXME: rdar://103979602
+StdFunctionTestSuite.test("FunctionIntToInt init from closure and call") {
+  let cClosure: @convention(c) (Int32) -> Int32 = { $0 + 1 }
+  let f = FunctionIntToInt(cClosure)
```
Comment (Xazax-hun) [inline test/Interop/Cxx/stdlib/use-std-function.swift:37]: Do we want to have a test case where the closure actually captures some local state? Or is it not yet supported?

Comment (egorzhdan) [reply test/Interop/Cxx/stdlib/use-std-function.swift:37]: That's not supported yet. `@convention(c)` closures (aka the ones that can be bridged to C function pointers) cannot capture context in Swift. In the future we'll need to add support for converting `@convention(thick)` closures to `std::function`, but it requires more work.

**Pull request merged**